### PR TITLE
fix: nan failure during training

### PR DESCRIPTION
**Pull request by ori-kron-wis**
Using SaveCheckpoint callback with on_exception can save the best optimal model up to the point it crashed due to Nan's in loss or gradients.
See an example (using Michal's data):

```
import scvi
from scvi.train._callbacks import SaveCheckpoint
from scvi.model import SCANVI
import pandas as pd
import numpy as np
import scanpy as sc
import torch
torch.set_float32_matmul_precision('high')

pd.set_option('display.max_rows', 50)
pd.set_option('display.max_columns', 50)
pd.set_option('display.width', 1000)
scvi.settings.seed = 0

early_stopping_kwargs = {
    'early_stopping': True,
    'early_stopping_monitor': 'elbo_validation', #'train_loss'
    'early_stopping_patience': 50,
    'early_stopping_mode': "min",
    'early_stopping_min_delta': 0.0,
    #'check_val_every_n_epoch': 1,
    #'check_finite': True,
}

ScviM = scvi.model.SCVI.load("/home/access/scvi_forScanVI4")

lvae = scvi.model.SCANVI.from_scvi_model(
                ScviM,
                unlabeled_category='unlabeled',
                labels_key="celltypes_steven2",
                linear_classifier=True,
            )
lvae.train(batch_size=1024,n_samples_per_label=100, max_epochs=500, gradient_clip_val=0,
           **early_stopping_kwargs , detect_anomaly=False, enable_checkpointing=True,
           callbacks=[SaveCheckpoint(monitor="elbo_validation", load_best_on_end=True)]) #breaks at epoch 58

#WE now want to laod this model and continue to train it
model = SCANVI.load("/home/access/.config/JetBrains/PyCharmCE2024.2/scratches/scvi_log/"
                    "2025-01-23_13-37-44_elbo_validation/"
                    "epoch=54-step=53295-elbo_validation=1255.7066650390625/",adata=ScviM.adata)
model.train(batch_size=2048,n_samples_per_label=50, max_epochs=500, gradient_clip_val=1,
           **early_stopping_kwargs , detect_anomaly=False, enable_checkpointing=True, plan_kwargs={"lr": 1e-2},
           callbacks=[SaveCheckpoint(monitor="elbo_validation", load_best_on_end=True)])

#running with detect_anomlay=True really slows down the whole thing
print("done")

```
We can then load it and continue training it (with or without parameters twicking)

### CODE REVIEW

File: src/scvi/train/_callbacks.py [hunk 1]
```diff
@@ -161,6 +166,51 @@ def on_train_end(self, trainer: pl.Trainer, pl_module: pl.LightningModule) -> No
...
+        """Save the model in case of unexpected exceptions, like Nan in loss or gradients"""
+        if not isinstance(exception, KeyboardInterrupt):
+            if self.loss_is_nan:
+                self.reason = (
+                    "\033[31m[Warning] NaN detected in the loss. Stopping training. "
+                    "Saving model....Please load it back and continue training\033[0m"
+                )
+            else:
+                self.reason = (
+                    "\033[31m[Warning] Exception occurred during training (Nan or Inf gradients). "
```
Comment (canergen) [inline src/scvi/train/_callbacks.py:185]: what's this \033[31m string?

File: src/scvi/train/_callbacks.py [hunk 2]
```diff
@@ -161,6 +166,51 @@ def on_train_end(self, trainer: pl.Trainer, pl_module: pl.LightningModule) -> No
...
+            # Loads the best model state into the model after the exception.
+            _, _, best_state_dict, _ = _load_saved_files(
+                self.best_model_path,
+                load_adata=False,
+                map_location=pl_module.module.device,
+            )
+            pyro_param_store = best_state_dict.pop("pyro_param_store", None)
+            pl_module.module.load_state_dict(best_state_dict)
+            if pyro_param_store is not None:
+                # For scArches shapes are changed and we don't want to overwrite
```
Comment (canergen) [inline src/scvi/train/_callbacks.py:199]: what's this comment here?

File: src/scvi/train/_callbacks.py [hunk 3]
```diff
@@ -161,6 +166,51 @@ def on_train_end(self, trainer: pl.Trainer, pl_module: pl.LightningModule) -> No
...
+                # these changed shapes.
+                pyro.get_param_store().set_state(pyro_param_store)
+            print(self.reason)
```
Comment (canergen) [inline src/scvi/train/_callbacks.py:202]: can we refine the printing here instead of two prints statements.

File: src/scvi/train/_trainer.py [hunk 4]
```diff
@@ -72,6 +72,8 @@ class Trainer(pl.Trainer):
         and in 'max' mode it will stop when the quantity monitored has stopped increasing.
     enable_progress_bar
         Whether to enable or disable the progress bar.
+    gradient_clip_val
```
Comment (canergen) [inline src/scvi/train/_trainer.py:75]: does it do anything?

File: tests/train/test_callbacks.py [hunk 5]
```diff
@@ -83,3 +84,31 @@ def test_model_cls(model_cls, adata: AnnData):
...
+    # adata.X[0,:] = 0
+
+    SCVI.setup_anndata(adata, batch_key="batch")
+
+    model = SCVI(adata)
+    model.train(max_epochs=5)
+
+    model.train(
+        max_epochs=5,
+        callbacks=[SaveCheckpoint(monitor="elbo_validation", load_best_on_end=True)],
```
Comment (canergen) [inline tests/train/test_callbacks.py:112]: does this give a NaN?

File: src/scvi/train/_callbacks.py (hunk 1 above)
Comment (ori-kron-wis) [reply src/scvi/train/_callbacks.py:185]: it prints the reason in red to screen

File: src/scvi/train/_callbacks.py (hunk 2 above)
Comment (ori-kron-wis) [reply src/scvi/train/_callbacks.py:199]: you tell me :-). it came from resolvi merge

File: src/scvi/train/_callbacks.py (hunk 3 above)
Comment (ori-kron-wis) [reply src/scvi/train/_callbacks.py:202]: we can do it in one line of course.

File: src/scvi/train/_trainer.py (hunk 4 above)
Comment (ori-kron-wis) [reply src/scvi/train/_trainer.py:75]: its working, but does it help in avoiding gradients nan exception? in my test case, no. 
it just changed the epoch it failed. Basically, its a common practice to use it when such a thing happens.

But, we can still use it via the trainer kwargs and not explicitly with the train function signature. Ill revert.

See velovi model, its part of it.

File: tests/train/test_callbacks.py (hunk 5 above)
Comment (ori-kron-wis) [reply tests/train/test_callbacks.py:112]: of course not. its a placeholder for now.
I was hoping that we could come up with a unit test data that casues the on exception to work in a pytest env. If we do we can use it here and check we get what we expect.
